src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Internal/ValidationStack.cs

```diff
@@ -0,0 +1,79 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using Microsoft.AspNetCore.Mvc.Internal;
+
+namespace Microsoft.AspNetCore.Mvc.ModelBinding.Internal
+{
+    internal class ValidationStack
+    {
+        // We tested the performance of a list at size 15 and found it still better than hashset, but to avoid a costly
+        // O(n) search at larger n we set the cutoff to 20. If someone finds the point where they intersect feel free to change this number.
+        protected virtual int CutOff
+        {
+            get { return 20; }
+        }
+
+        protected readonly List<object> _list = new List<object>();
+        protected HashSet<object> _hashSet;
```

Comment:
Fields before properties or make these properties as well. Not normal to expose fields.

Also, use `internal` instead of `protected`? `protected` is too much like `public` and customers shouldn't derive from this `Internal` class.